generator/readers.ts

```diff
@@ -0,0 +1,150 @@
+import * as yamlFrontMatter from 'front-matter';
+import * as yaml from 'js-yaml';
+import * as path from 'path';
+import * as slug from 'slug';
+
+import {GuideHTMLFileWithMetadata, LearningPath, TopLevelFile} from './file-types.js';
+import * as fs from './fsp.js';
+
+const REQUIRED_ATTRIBUTES = [
+  'page_type', 'title', 'author', 'description', 'web_updated_on',
+  'web_published_on'
+];
+
+const GUIDE_REQUIRED_ATTRIBUTES = [...REQUIRED_ATTRIBUTES, 'web_lighthouse'];
+const CODELAB_REQUIRED_ATTRIBUTES = [...REQUIRED_ATTRIBUTES, 'glitch'];
+const GUIDE_CONFIGURATION_REQUIRED_ATTRIBUTES =
+    ['title', 'description', 'overview', 'topics', 'order'];
+
+async function readYamlAndAssertAttributes(
+    requiredAttributes: string[], fileName: string) {
+  const {attributes, body} =
+      yamlFrontMatter(await fs.readFile(fileName, 'utf8'));
+
+  for (const requiredAttribute of requiredAttributes) {
```

Comment:
It normally does, but because we are dynamically reading here, TS has no clue what you read from disk. Therefore this function exists to assert it is in the correct shape, thus afterwards we do adhere to the interface (and TS will check for that).